parsley-debug/shared/src/main/scala/parsley/debugger/ParseAttempt.scala

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright 2020 Parsley Contributors <https://github.com/j-mie6/Parsley/graphs/contributors>
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+package parsley.debugger
+
+import ParseAttempt._ // scalastyle:ignore underscore.import
+
+/** A representation of the attempts a parser has made during parse-time.
+  *
+  * @since 4.5.0
+  */
+final class ParseAttempt private [parsley] (
+    inp: Input,
+    fof: Offset,
+    tof: Offset,
+    fps: Pos,
+    tps: Pos,
+    scs: Success,
+    res: Result
+) {
+    /** The input parsed, as raw text. */
+    val rawInput: Input = inp
+
+    /** This offset is where the parse attempt started in the input. */
+    val fromOffset: Offset = fof
+
+    /** This offset is where the parse attempt finished in the input. */
+    val toOffset: Offset = tof
+
+    /** [[fromOffset]] represented as a (line, column) pair. */
+    val fromPos: Pos = fps
+
+    /** [[toOffset]] represented as a (line, column pair). */
+    val toPos: Pos = tps
+
+    /** Was this parse attempt successful?
+      *
+      * @note [[success]] if and only if [[result]] is defined (contains a value).
+      */
+    val success: Success = scs
+
+    /** If this parse attempt was successful, what did it return? It is guaranteed that `result.isDefined` is true
+      * if and only if the attempt is successful.
+      *
+      * @note [[success]] if and only if [[result]] is defined (contains a value).
+      */
+    val result: Result = res
+
+    // Make sure this class has not been used improperly.
+    assert(success == result.isDefined)
+
+    // Utility copy method only to be used internally.
+    private [parsley] def copy(
+        inp: Input = rawInput,
+        fof: Offset = fromOffset,
+        tof: Offset = toOffset,
+        fps: Pos = fromPos,
+        tps: Pos = toPos,
+        scs: Success = success,
+        res: Result = result
+    ): ParseAttempt =
+        new ParseAttempt(inp, fof, tof, fps, tps, scs, res)
+}
+
+object ParseAttempt {
```

Comment:
Needs to be private up to parsley. Can't expose this user-facing or it's a bin-compat problem.

Reply:
The object? Would the `unapply` still get called even if it's private?

Reply:
In internal code, yes. Not in public code: because `unapply` can then never change return type, and we have binary lock in

Reply:
Hm, ideally matching should work in public code for `ParseAttempt`, but I see why that could introduce some issues.

Reply:
Yeah, we must avoid binary incompatibility at _all_ costs